core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -64,18 +66,54 @@
 
 	private static final Logger logger = LoggerFactory.getLogger(Transaction.class);
 
+	/**
+	 * The period during the {@link #close()} method between waiting for
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)}, logging that the transaction has not
+	 * completed, and trying again up to {@link #MAX_CLOSE_WAIT_PERIODS} attempts.
+	 */
+	private static final int EXECUTOR_SHUTDOWN_WAIT_PERIOD = 10000;
+
+	/**
+	 * The maximum number of times during the {@link #close()} method that
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)} will be called before giving up and calling
+	 * {@link ExecutorService#shutdownNow()}.
+	 */
+	private static final int MAX_CLOSE_WAIT_PERIODS = 3;
+
+	/**
+	 * Set to true when entering the {@link #close()} method for the first time, to ensure that only a single
+	 * thread executes the close operations.
+	 */
 	private final AtomicBoolean isClosed = new AtomicBoolean(false);
 
+	/**
+	 * Set to true when the {@link #close()} method is about to complete for the first invocation.
+	 */
+	private final AtomicBoolean closeCompleted = new AtomicBoolean(false);
+
 	private final UUID id;
 
 	private final Repository rep;
 
 	private final RepositoryConnection txnConnection;
 
+	/**
+	 * The {@link ExecutorService} that performs all of the operations related to this Transaction.
+	 */
 	private final ExecutorService executor = Executors.newSingleThreadExecutor(
 			new ThreadFactoryBuilder().setNameFormat("rdf4j-transaction-%d").build());
 
-	private final List<Future<?>> futures = new ArrayList<>();
+	/**
+	 * An append-only {@link Queue} that tracks the state of all of the futures involved in this Transaction.
+	 */
+	private final Queue<Future<?>> futures = new LinkedBlockingQueue<>();
+
+	/**
+	 * This {@link Semaphore} is used to ensure that only a single task is executed by the ExecutorService,
+	 * and that these tasks are not run while {@link #hasActiveOperations()} is determining whether there are
+	 * outstanding tasks open for this Transaction.
+	 */
+	private final Semaphore semaphore = new Semaphore(1);
```

Comment:
I still haven't quite gotten my head around the issue. Let me talk my way through this: 

In the previous setup (synchronized blocks, no semaphore), when a new operation was invoked on the transaction, this locked (via `synchronized`) on the futures list. An operation was scheduled in the executor and its future added to the list, after which the lock is released. `hasActiveOperations` itself also locks on `futures` access. Which means that while `hasActiveOperations` is running, no other threads can add new operations to the transaction. Correct sofar? 

Potential problem 1: `hasActiveOperations` loops over `Future` objects that may concurrently change their state to `done`. It iterates until it finds at least one future that _isn't_ done. So the question is: is there a problem when a future gets adapted during looping? As far I can tell, there isn't: if it checks whether Future `x` is done and it gets a `false` (and immediately after that `x` is completed), it just continues iterating. Worst case is that it mistakenly returns `true` because it believes `x` is still active, while in reality it's already completed. So a false positive, which is no problem from a cleanup perspective.

Potential problem 2: `hasActiveOperations` loops and finds no active operations, however concurrently another thread is waiting to obtain a lock to schedule a new operation and add it to the futures list. `hasActiveOperations` returns `false`, the registry closes and removes the transaction, despite the fact that a new operation was in fact received. This sounds dire, but I think it can't actually happen: every incoming transaction operation first obtains the Transaction object from the registry. So it either obtains the Transaction object (which means the timer is reset, so it won't activate the registry cleanup for that object and therefore won't invoke `hasActiveOperations`), or the `Transaction` has already been cleaned up, and is therefore no longer accessible from the cache, resulting in an _expected_ timeout error. 

Sorry to ramble. Do you see any flaws in my thinking? I'd be happy to add the semaphore stuff in, but I am slightly wary of over-complicating the locking logic without good cause, so I want to make sure I understand _why_ this is necessary :)